File: Scripts/bench_encode.py
#!/usr/bin/env python3
"""
Питоновская половина замера кодирования. Swift-половина — `rerank-run --bench`.

Обе читают ОДИН файл входа (`bench_make_input.py`) и делают ровно одно и то
же: токенизируют префиксы, гоняют их пачками через базу, продолжают хвостами
запросов, отбирают читаемые головой слои.

Что засекается по отдельности и почему:

  load        — веса и словарь. В разрыв «Swift против Python» не входит и
                вынесено, чтобы не входило.
  warmup      — первые батчи ОТБРАСЫВАЮТСЯ. На Metal первый запуск каждой
                новой формы тянет за собой сборку конвейера, и без прогрева
                этот разовый расход размазывается по замеру, притворяясь
                пропускной способностью.
  tokenize    — чистый CPU, без GPU вовсе.
  gpu_prefix  — batch_ids + проход базы + eval (синхронизация обязательна,
                иначе засекается постановка в очередь, а не работа).
  gpu_tail    — то же для хвостов поверх состояния префикса.
  select      — отбор слоёв и вынос в numpy.

Ряд по батчам печатается целиком: разовый расход виден только по форме
кривой, суммой его не отличить от медленного кода.

    .venv/bin/python bench_encode.py --model world_0.1b_x070.safetensors \
        --input /tmp/bench_input.json --warmup 3
"""
import argparse
import json
import os
import resource
import subprocess
import sys
import time

import mlx.core as mx
import numpy as np

# Рабочее ДЕРЕВО, а не установленная копия. Скрипт лежит в SwiftRWKV/Scripts,
# поэтому sys.path[0] указывает туда, и `import rwkv_metal` без этой строки
# подхватывает site-packages — то есть меряет не тот код, который лежит
# рядом с замером. Молчаливо: версии обычно близкие, а числа разные.
_REPO = os.environ.get("RWKV_METAL_REPO",
                       os.path.expanduser("~/Develop/rwkv-metal"))
sys.path.insert(0, _REPO)


# Память — ПО ДАННЫМ СИСТЕМЫ, а не по счётчикам MLX. `mx.get_peak_memory()`
# знает только про свой пул: ни про numpy-буферы, ни про веса, ни про то,
# ушла ли машина в своп. А если процесс свопится, врут и замеры времени.
def rss_gb() -> float:
    out = subprocess.run(["ps", "-o", "rss=", "-p", str(os.getpid())],
                         capture_output=True, text=True).stdout.strip()
    return int(out) * 1024 / 1e9 if out else float("nan")


def peak_rss_gb() -> float:
    raw = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    return (raw if sys.platform == "darwin" else raw * 1024) / 1e9


def footprint_gb() -> float:
    """Physical footprint: буферы Metal живут в IOAccelerator и в RSS
    попадают неполностью."""
    try:
        out = subprocess.run(["vmmap", "--summary", str(os.getpid())],
                             capture_output=True, text=True).stdout
        for line in out.splitlines():
            if "Physical footprint:" in line and "Peak" not in line:
                v = line.split(":")[1].strip()
                n = float(v.rstrip("GMK"))
                return n if v.endswith("G") else n / 1024 if v.endswith("M") else n / 1e6
    except Exception:
        pass
    return float("nan")


def swap_counters():
    try:
        vm = subprocess.run(["vm_stat"], capture_output=True, text=True).stdout
        ins = outs = 0
        for line in vm.splitlines():
            if "Swapins" in line:
                ins = int(line.split(":")[1].strip().rstrip("."))
            elif "Swapouts" in line:
                outs = int(line.split(":")[1].strip().rstrip("."))
        return ins, outs
    except Exception:
        return 0, 0


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--model", required=True)
    ap.add_argument("--input", required=True)
    ap.add_argument("--warmup", type=int, default=3,
                    help="сколько первых пачек префиксов отбросить")
    ap.add_argument("--layers", type=int, default=5)
    ap.add_argument("--out", default=None, help="куда положить JSON с числами")
    ap.add_argument("--cache-limit", type=float, default=2.0,
                    help="потолок буферного кэша MLX, ГБ; 0 — не трогать")
    a = ap.parse_args()

    with open(a.input, encoding="utf-8") as f:
        inp = json.load(f)

    # Потолок буферного кэша Metal. БЕЗ него замер недействителен: формы
    # батчей плавают, переиспользовать буферы нечего, кэш растёт линейно по
    # числу пачек. Замерено: на восьмистах префиксах footprint дошёл до
    # 11.6 ГБ (RSS при этом показывал 0.4 — буферы живут в IOAccelerator), и
    # машина ушла в своп.
    #
    # Этого потолка нет в самом `encode_pairs` — то есть дефект здесь не
    # только у замера. В Swift он вылечен (`RerankEncodeConfig.cacheLimitGB`
    # плюс округление длины батча), в Python пока нет.
    if a.cache_limit > 0:
        mx.set_cache_limit(int(a.cache_limit * 1e9))

    swap0 = swap_counters()
    t = time.perf_counter()
    import rwkv_metal
    from rwkv_metal.model.rwkv7_x070 import RWKV7X070, lora_ranks
    from rwkv_metal.tokenizer import WorldTokenizer
    from rwkv_metal.reranker import Reranker, RerankerConfig
    from rwkv_metal.reranker.encode import (_batch_ids, _encode_prefix_ids,
                                            _encode_suffix_ids)
    from rwkv_metal.reranker.data import PairTemplate
    import_s = time.perf_counter() - t

    t = time.perf_counter()
    # Геометрия — из форм весов, как в dump_block_reference.py: конфигурации,
    # которую можно задать неверно, тут просто нет.
    weights = mx.load(os.path.expanduser(a.model))
    n_layer = max(int(k.split(".")[1])
                  for k in weights if k.startswith("blocks.")) + 1

    class Cfg:
        pass

    cfg = Cfg()
    cfg.n_layer = n_layer
    cfg.n_embd = weights["ln_out.weight"].shape[0]
    cfg.vocab_size = weights["head.weight"].shape[0]
    cfg.head_size = weights["blocks.0.tmix.k_k"].shape[1]
    cfg.n_head = cfg.n_embd // cfg.head_size
    base = RWKV7X070(cfg, lora_ranks(cfg.n_embd))
    base.load_weights(list(weights.items()))
    mx.eval(base.parameters())
    tok = WorldTokenizer()
    model = Reranker(base, RerankerConfig(layer_idx=(a.layers,)))
    load_s = time.perf_counter() - t
    print(f"модель: L={cfg.n_layer} D={cfg.n_embd} V={cfg.vocab_size}, "
          f"rwkv_metal из {_REPO}", flush=True)

    template = PairTemplate()
    prefixes = inp["prefixes"]
    doc_batch, query_batch = inp["doc_batch"], inp["query_batch"]

    def prefix_ids(doc):
        return _encode_prefix_ids(tok, template, inp["instruct"], doc,
                                  inp["max_doc_tokens"])

    def suffix_ids(doc, query):
        return _encode_suffix_ids(tok, template, doc, query,
                                  inp["max_query_tokens"],
                                  inp["max_doc_tokens"], inp["terminator"])

    acc = {"tokenize": 0.0, "gpu_prefix": 0.0, "gpu_tail": 0.0, "select": 0.0}
    series = []
    n_prefix_done = n_tail_done = 0
    shapes_prefix, shapes_tail = set(), set()

    wall0 = time.perf_counter()
    for bi, start in enumerate(range(0, len(prefixes), doc_batch)):
        chunk = prefixes[start:start + doc_batch]
        warm = bi < a.warmup
        b = {}

        t = time.perf_counter()
        seqs = [prefix_ids(p["doc"]) for p in chunk]
        b["tokenize"] = time.perf_counter() - t

        t = time.perf_counter()
        idx, mask, end_idx = _batch_ids(seqs)
        st = model.encode(idx, mask=mask, end_idx=end_idx)
        st.eval()
        b["gpu_prefix"] = time.perf_counter() - t
        shapes_prefix.add(tuple(idx.shape))

        jobs = [(local, q) for local, p in enumerate(chunk)
                for q in p["queries"]]
        b["gpu_tail"] = b["select"] = b["tok_tail"] = 0.0
        for qs in range(0, len(jobs), query_batch):
            part = jobs[qs:qs + query_batch]
            t = time.perf_counter()
            qseqs = [suffix_ids(chunk[l]["doc"], q) for l, q in part]
            b["tok_tail"] += time.perf_counter() - t

            t = time.perf_counter()
            locs = mx.array(np.array([p[0] for p in part], dtype=np.int32))
            sub = st[locs]
            qidx, qmask, qend = _batch_ids(qseqs)
            st_pair = model.encode(qidx, mask=qmask, end_idx=qend, state=sub)
            sel = model.select(st_pair)
            mx.eval(sel)
            b["gpu_tail"] += time.perf_counter() - t
            shapes_tail.add(tuple(qidx.shape))

            t = time.perf_counter()
            _ = np.array(sel.astype(mx.float32)).astype(np.float16)
            b["select"] += time.perf_counter() - t

        b["tokenize"] += b.pop("tok_tail")
        if not warm:
            for k in acc:
                acc[k] += b[k]
            n_prefix_done += len(chunk)
            n_tail_done += len(jobs)
        series.append({"batch": bi, "warmup": warm,
                       "prefixes": len(chunk), "tails": len(jobs),
                       "rss_gb": round(rss_gb(), 2),
                       **{k: round(v * 1000, 2) for k, v in b.items()}})
    wall = time.perf_counter() - wall0
    swap1 = swap_counters()

    measured = sum(acc.values())
    swapped_out = swap1[1] - swap0[1]
    res = {
        "side": "python", "cache_limit_gb": a.cache_limit,
        "valid": swapped_out == 0, "import_s": import_s, "load_s": load_s,
        "wall_s": wall, "measured_s": measured,
        "prefixes": n_prefix_done, "tails": n_tail_done,
        "warmup_batches": a.warmup,
        "ms_per_prefix": acc["gpu_prefix"] / max(1, n_prefix_done) * 1000,
        "ms_per_tail": acc["gpu_tail"] / max(1, n_tail_done) * 1000,
        "distinct_prefix_shapes": len(shapes_prefix),
        "distinct_tail_shapes": len(shapes_tail),
        "rss_gb": rss_gb(), "peak_rss_gb": peak_rss_gb(),
        "footprint_gb": footprint_gb(),
        "swapins": swap1[0] - swap0[0], "swapouts": swapped_out,
        "phases_s": acc, "series": series,
    }
    print(json.dumps({k: v for k, v in res.items() if k != "series"},
                     ensure_ascii=False, indent=2))
    if swapped_out > 0:
        print(f"ЗАМЕР НЕДЕЙСТВИТЕЛЕН: {swapped_out} страниц ушло в своп. "
              "Числа выше меряют диск, а не код.")
    if a.out:
        with open(a.out, "w", encoding="utf-8") as f:
            json.dump(res, f, ensure_ascii=False, indent=2)


if __name__ == "__main__":
    main()

File: Sources/RerankRun/BenchEncode.swift
import Foundation
import MLX
import RWKVGen
import RWKVRerank

// ───────────────────────────────────────────────────────────────────────
//  Swift-половина замера кодирования. Питоновская — Scripts/bench_encode.py.
//
//  Обе читают ОДИН файл входа (Scripts/bench_make_input.py) и делают ровно
//  одно и то же. Это единственный способ сравнивать реализации, а не
//  постановки: разные срезы данных дают разные длины документов, разное
//  число уникальных префиксов и разное число пачек, и любая разница во
//  времени объясняется тогда чем угодно.
//
//  Что засекается по отдельности и почему:
//
//    load        — веса и словарь. В разрыв не входит и вынесено, чтобы
//                  не входило.
//    warmup      — первые пачки ОТБРАСЫВАЮТСЯ. На Metal первый запуск
//                  каждой новой формы тянет за собой сборку конвейера;
//                  без прогрева этот разовый расход размазывается по
//                  замеру и притворяется пропускной способностью.
//    tokenize    — чистый CPU, GPU не трогается вовсе.
//    gpuPrefix   — batchIds + проход базы + eval. Синхронизация
//                  ОБЯЗАТЕЛЬНА: без неё засекается постановка в очередь,
//                  а не работа, и Swift «выигрывает» на пустом месте.
//    gpuTail     — то же для хвостов поверх состояния префикса.
//    select      — отбор слоёв и вынос в CPU-буфер.
//
//  Ряд по пачкам печатается целиком: разовый расход виден только по форме
//  кривой, суммой его от медленного кода не отличить.
// ───────────────────────────────────────────────────────────────────────

struct BenchInput: Decodable {
    struct Prefix: Decodable { var doc: String; var queries: [String] }
    var instruct: String
    var max_doc_tokens: Int
    var max_query_tokens: Int
    var terminator: Int?
    var doc_batch: Int
    var query_batch: Int
    var prefixes: [Prefix]
}

/// RSS процесса, ГБ — тем же способом, каким его видит система.
func benchRssGB() -> Double {
    let p = Process()
    p.executableURL = URL(fileURLWithPath: "/bin/ps")
    p.arguments = ["-o", "rss=", "-p", String(ProcessInfo.processInfo.processIdentifier)]
    let pipe = Pipe(); p.standardOutput = pipe
    guard (try? p.run()) != nil else { return .nan }
    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    p.waitUntilExit()
    let s = String(data: data, encoding: .utf8)?
        .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return Double(s).map { $0 * 1024 / 1e9 } ?? .nan
}

/// Счётчики свопа системы. Замер, снятый во время свопа, — не «чуть хуже»,
/// а недействительный: страницы ходят на диск, и меряется он, а не код.
func swapCounters() -> (Int, Int) {
    let p = Process()
    p.executableURL = URL(fileURLWithPath: "/usr/bin/vm_stat")
    let pipe = Pipe(); p.standardOutput = pipe
    guard (try? p.run()) != nil else { return (0, 0) }
    let out = String(data: pipe.fileHandleForReading.readDataToEndOfFile(),
                     encoding: .utf8) ?? ""
    p.waitUntilExit()
    var ins = 0, outs = 0
    for line in out.split(separator: "\n") {
        let v = Int(line.split(separator: ":").last?
            .trimmingCharacters(in: CharacterSet(charactersIn: " .")) ?? "") ?? 0
        if line.contains("Swapins") { ins = v }
        if line.contains("Swapouts") { outs = v }
    }
    return (ins, outs)
}

func runBench(inputPath: String, modelPath: String, vocabPath: String,
              warmup: Int, layers: [Int], bucket: Int, minT: Int,
              cacheLimitGB: Double, outPath: String?) throws {

    // Потолок буферного кэша Metal. БЕЗ него замер недействителен: формы
    // батчей плавают, переиспользовать буферы нечего, кэш растёт линейно по
    // числу пачек и уводит машину в своп — замерено, 10+ ГБ на восьмистах
    // префиксах. Штатный `encodePairs` его ставит, и замер обязан ставить
    // тот же, иначе меряется не тот код, который работает в продакшене.
    if cacheLimitGB > 0 {
        MLX.GPU.set(cacheLimit: Int(cacheLimitGB * 1e9))
    }
    let swap0 = swapCounters()

    let raw = try Data(contentsOf: URL(fileURLWithPath: expand(inputPath)))
    let inp = try JSONDecoder().decode(BenchInput.self, from: raw)

    var t = Date()
    let weights = try loadArrays(url: URL(fileURLWithPath: expand(modelPath)))
    let nLayer = weights.keys.compactMap { k -> Int? in
        guard k.hasPrefix("blocks.") else { return nil }
        return Int(k.split(separator: ".")[1])
    }.max().map { $0 + 1 } ?? 0
    let cfg = X070Config(nLayer: nLayer,
                         nEmbd: weights["ln_out.weight"]!.shape[0],
                         headSize: weights["blocks.0.tmix.k_k"]!.shape[1],
                         vocab: weights["head.weight"]!.shape[0])
    let backbone = X070Backbone(weights: weights, cfg: cfg)
    guard let tok = WorldTokenizer(
            vocabURL: URL(fileURLWithPath: expand(vocabPath))) else {
        print("не удалось прочитать словарь"); exit(2)
    }
    let model = try Reranker(base: backbone,
                             cfg: RerankerConfig(layerIdx: layers))
    let loadS = Date().timeIntervalSince(t)

    let encCfg = RerankEncodeConfig(
        maxDocTokens: inp.max_doc_tokens, maxQueryTokens: inp.max_query_tokens,
        terminator: inp.terminator, docBatch: inp.doc_batch,
        queryBatch: inp.query_batch, lengthBucket: bucket,
        lengthBucketMinTokens: minT)

    var acc = ["tokenize": 0.0, "gpu_prefix": 0.0, "gpu_tail": 0.0,
               "select": 0.0]
    var series: [[String: Any]] = []
    var nPrefix = 0, nTail = 0
    var shapesPrefix = Set<[Int]>(), shapesTail = Set<[Int]>()

    let wall0 = Date()
    var bi = 0
    var start = 0
    while start < inp.prefixes.count {
        let end = Swift.min(start + inp.doc_batch, inp.prefixes.count)
        let chunk = Array(inp.prefixes[start ..< end])
        let warm = bi < warmup
        var b = ["tokenize": 0.0, "gpu_prefix": 0.0, "gpu_tail": 0.0,
                 "select": 0.0]

        t = Date()
        let seqs = chunk.map {
            RerankEncoder.prefixIds(tok, encCfg, instruct: inp.instruct,
                                    document: $0.doc)
        }
        b["tokenize"]! += Date().timeIntervalSince(t)

        t = Date()
        let (idx, mask, endIdx) = RerankEncoder.batchIds(seqs, bucket: bucket,
                                                        minT: minT)
        let st = model.encode(idx, mask: mask, endIdx: endIdx)
        st.evaluated()
        b["gpu_prefix"]! += Date().timeIntervalSince(t)
        shapesPrefix.insert(idx.shape)

        var jobs: [(local: Int, query: String)] = []
        for (local, p) in chunk.enumerated() {
            for q in p.queries { jobs.append((local, q)) }
        }

        var qs = 0
        while qs < jobs.count {
            let qe = Swift.min(qs + inp.query_batch, jobs.count)
            let part = Array(jobs[qs ..< qe])

            t = Date()
            let qseqs = part.map {
                RerankEncoder.suffixIds(tok, encCfg,
                                        document: chunk[$0.local].doc,
                                        query: $0.query)
            }
            b["tokenize"]! += Date().timeIntervalSince(t)

            t = Date()
            let sub = st.gather(part.map { $0.local })
            let (qidx, qmask, qend) = RerankEncoder.batchIds(qseqs, bucket: bucket,
                                                            minT: minT)
            let pairState = model.encode(qidx, mask: qmask, endIdx: qend,
                                         state: sub)
            let sel = model.select(pairState)
            eval(sel)
            b["gpu_tail"]! += Date().timeIntervalSince(t)
            shapesTail.insert(qidx.shape)

            t = Date()
            // Тот же вынос в CPU-буфер, что делает питоновская сторона:
            // fp32 → fp16 → байты. Без него замер кончался бы на GPU, а
            // кэш пишется всё-таки на хост.
            _ = sel.asType(.float32).asType(.float16).asData(disambiguate: true)
            b["select"]! += Date().timeIntervalSince(t)

            qs = qe
        }

        if !warm {
            for (k, v) in b { acc[k]! += v }
            nPrefix += chunk.count
            nTail += jobs.count
        }
        var row: [String: Any] = ["batch": bi, "warmup": warm,
                                  "prefixes": chunk.count, "tails": jobs.count,
                                  "rss_gb": (benchRssGB() * 100).rounded() / 100]
        for (k, v) in b { row[k] = (v * 1000 * 100).rounded() / 100 }
        series.append(row)
        bi += 1
        start = end
    }
    let wall = Date().timeIntervalSince(wall0)
    let swap1 = swapCounters()
    let swappedOut = swap1.1 - swap0.1

    let res: [String: Any] = [
        "swapouts": swappedOut, "swapins": swap1.0 - swap0.0,
        "cache_limit_gb": cacheLimitGB,
        "valid": swappedOut == 0,
        "side": "swift", "load_s": loadS, "wall_s": wall,
        "measured_s": acc.values.reduce(0, +),
        "prefixes": nPrefix, "tails": nTail,
        "warmup_batches": warmup, "length_bucket": bucket,
        "length_bucket_min_t": minT,
        "ms_per_prefix": acc["gpu_prefix"]! / Double(Swift.max(1, nPrefix)) * 1000,
        "ms_per_tail": acc["gpu_tail"]! / Double(Swift.max(1, nTail)) * 1000,
        "distinct_prefix_shapes": shapesPrefix.count,
        "distinct_tail_shapes": shapesTail.count,
        "rss_gb": benchRssGB(), "footprint_gb": processFootprintGB(),
        "phases_s": acc, "series": series,
    ]
    var head = res
    head.removeValue(forKey: "series")
    if let d = try? JSONSerialization.data(withJSONObject: head,
                                           options: [.prettyPrinted, .sortedKeys]),
       let s = String(data: d, encoding: .utf8) { print(s) }
    if swappedOut > 0 {
        print("ЗАМЕР НЕДЕЙСТВИТЕЛЕН: \(swappedOut) страниц ушло в своп. "
              + "Числа выше меряют диск, а не код. Уменьши вход или "
              + "потолок кэша.")
    }
    if let outPath {
        let url = URL(fileURLWithPath: expand(outPath))
        try? FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        if let d = try? JSONSerialization.data(withJSONObject: res,
                                               options: [.prettyPrinted, .sortedKeys]) {
            try? d.write(to: url)
        }
    }
}
